File: client_python/src/Transaction.py
from src.service.Session.TransactionService import TransactionService

class Transaction(object):
    """ Presents the Grakn interface to the user, actual work with GRPC happens in TransactionService """

    def __init__(self, transaction_service: TransactionService):
        self._tx_service = transaction_service


    def query(self, query: str):
        return self._tx_service.query(query)

    def commit(self):
        self._tx_service.commit()
        self.close()

    def close(self):
        self._tx_service.close() # close the service

    def get_concept(self, concept_id: str): # TODO annotate return types
        return self._tx_service.get_concept(concept_id)

    def get_schema_concept(self, label: str): 
        return self._tx_service.get_schema_concept(label)

    def get_attributes_by_value(self, attribute_value, data_type):
        return self._tx_service.get_attributes_by_value(attribute_value, data_type)

    def put_entity_type(self, label: str):
        return self._tx_service.put_entity_type(label)

    def put_relationship_type(self, label: str):
        return self._tx_service.put_relationship_type(label)

    def put_attribute_type(self, label: str, data_type):
        return self._tx_service.put_attribute_type(label, data_type)

    def put_role(self, label: str):
        return self._tx_service.put_role(label)

    def put_rule(self, label: str, when: str, then: str):
        return self._tx_service.put_rule(label, when, then)

File: client-nodejs/src/Transaction.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/**
 * This object represents a transaction which is produced by Grakn and allows the user to construct and perform
 * basic look ups to the knowledge base. This also allows the execution of Graql queries.
 * 
 * @param {Object} txService Object implementing all the functionalities of gRPC Transaction service as defined in grakn.proto
 */
function Transaction(txService) {
    this.txService = txService;
}

/**
 * Executes a given Graql query on the keyspace this transaction is bound to
 * @param {String} query String representing a Graql query 
 */
Transaction.prototype.query = function executeQuery(query, options) {
    return this.txService.query(query, options);
};

/**
 * Commits any changes to the graph and closes the transaction. The user must use the Session object to
 * get a new open transaction.
 */
Transaction.prototype.commit = async function () {
    await this.txService.commit();
    return this.close();
}

/**
 * Get the Concept with identifier provided, if it exists.
 * 
 * @param {String} conceptId A unique identifier for the Concept in the graph.
 */
Transaction.prototype.getConcept = function (conceptId) {
    return this.txService.getConcept(conceptId);
}

Transaction.prototype.getSchemaConcept = function (label) {
    return this.txService.getSchemaConcept(label);
}

Transaction.prototype.getAttributesByValue = function (attributeValue, dataType) {
    return this.txService.getAttributesByValue(attributeValue, dataType);
}

Transaction.prototype.putEntityType = function (label) {
    return this.txService.putEntityType(label);
}

Transaction.prototype.putRelationshipType = function (label) {
    return this.txService.putRelationshipType(label);
}

Transaction.prototype.putAttributeType = function (label, dataType) {
    return this.txService.putAttributeType(label, dataType);
}

Transaction.prototype.putRole = function (label) {
    return this.txService.putRole(label);
}

Transaction.prototype.putRule = function (label, when, then) {
    return this.txService.putRule(label, when, then);
}

Transaction.prototype.close = function () {
    return this.txService.close();
}

module.exports = Transaction;